### ISSUE TITLE
No CRUD generation support for ARRAY types.

### ISSUE THREAD

**Issue by gmos** [piccolo-orm/piccolo_api#109]
At version 0.29.0 there seems to be no support for ARRAY types in the CRUD generation.
Code is generated, but an ARRAY of VARCHAR comes out as a single Varchar, and the field cannot be properly used.

In the Piccolo Admin the  Arrays of Varchar are handled ok.   Are there already plans to support Array types in CRUD?

**Comment by dantownsend** [comment]
@gmos Thanks for reporting this.

``PiccoloCRUD`` should support arrays - there could be a bug somewhere.

In the example below, the `years_nominated` field is an array. The API should render it like a list.

<img width="457" alt="Screenshot 2021-12-02 at 10 05 43" src="https://user-images.githubusercontent.com/350976/144400846-51c6be7b-4b7e-4cf7-b88b-63c9e40eeea8.png">

Is the problem when trying to post a new value back?

**Comment by sinisaos** [comment]
@gmos @dantownsend `POST` also work as expected.

![post](https://user-images.githubusercontent.com/30960668/144437430-9fad8e92-0aeb-47fc-831d-3f2833833e39.png)

**Comment by dantownsend** [comment]
@sinisaos Thanks for checking that.

**Comment by gmos** [comment]
Screenshots are from the swagger .../docs page.  I expanded the root GET.

Parameters docs:
![image](https://user-images.githubusercontent.com/1526544/144451030-4b712d0b-b760-4171-bb69-da1cf4bde8fe.png)
Here it goes wrong.  No array structure to be found.  The sensors field is depicted as a simple Varchar.

And this is when I make the actual GET call:
![image](https://user-images.githubusercontent.com/1526544/144452588-6bc9cc00-d2cf-4a58-8ba4-448aa05a1448.png)

As you can see the response schema is good.  the issue is about the generated parameter input for the root GET.

**Comment by dantownsend** [comment]
@gmos Cool, thanks - that's really useful. We'll investigate.

**Comment by gmos** [comment]
Here is the column definition from the Table subclass:
```
    sensors = Array(
        base_column=Varchar(
            length=16,
            null=False,
        )
    )
```
And here the automatic creation of the lot.
```
hoses_router = APIRouter()
hoses_crud = PiccoloCRUD(TempHoses, read_only=False)

FastAPIWrapper(
    "/hoses",
    fastapi_app=cast(FastAPI, hoses_router),  # TODO Get away with the cast
    piccolo_crud=hoses_crud,
    fastapi_kwargs=FastAPIKwargs(
        all_routes={"tags": ["TempHoses"]},
    ),
)
```

Side question: why do I need a cast to get rid of the type warning?  AM I doing something wrong assigning a PiccoloCRUD instance to the FastAPIwrapper?

**Comment by dantownsend** [comment]
@gmos The type error might be because you're using `FastAPIWrapper` with a `APIRouter` instance, rather than a `FastAPI` instance.

We probably just need to update this type annotation to be `t.Union[FastAPI, APIRouter]`:

https://github.com/piccolo-orm/piccolo_api/blob/e4eb74c9b0d8db961287fe9ffc37cb4576f4824c/piccolo_api/fastapi/endpoints.py#L94

**Comment by sinisaos** [comment]
@dantownsend Problem is when you use `openapi` request url is `http://localhost:8000/posts/?tags=tag` (which raise Pydantic ValidationError `value is not a valid list (type=type_error.list)`) and it should be `http://localhost:8000/posts/?tags[]=tag` like in Piccolo Admin or if you pass request url directly in browser to get correct result.

**Comment by dantownsend** [comment]
@sinisaos Well remembered - I forgot about that.

I wonder if we can add a custom validator to the Pydantic model:

https://pydantic-docs.helpmanual.io/usage/validators/

**Comment by sinisaos** [comment]
@dantownsend I think I found a problem in `openapi` filters. We need to use `ModeField.outer_type_` https://github.com/piccolo-orm/piccolo_api/blob/8b66ff3665f506a4eeeb3cd6259ff24cdea5beb9/piccolo_api/fastapi/endpoints.py#L401 because `ModeField.type_` shows the wrong type (class `int` or `str` insted `List[int]` or `List[str]`). 
We need to change the `line 401` to this `type_ = _field.outer_type_` and the result is.

![arrays](https://user-images.githubusercontent.com/30960668/145349102-c76be263-ca45-4362-81e5-fc0df7848e76.png)

Then I add an extra check in `crud endpoints _parse_params` method for the array columns so that `openapi request url` accepts multiple query params without square brackets. Method now looks like this

```python
def _parse_params(self, params: QueryParams) -> t.Dict[str, t.Any]:
  params_map: t.Dict[str, t.Any] = {
      i[0]: [j[1] for j in i[1]]
      for i in itertools.groupby(params.multi_items(), lambda x: x[0])
  }

  array_columns = [
      i._meta.name
      for i in self.table._meta.columns
      if i.value_type == list
  ]

  output = {}

  for key, value in params_map.items():
      if key.endswith("[]") or key.rstrip("[]") in array_columns:
          # Is either an array, or multiple values have been passed in
          # for another field.
          key = key.rstrip("[]")
      elif len(value) == 1:
          value = value[0]

      output[key] = value

  return output
```
And result is

![query](https://user-images.githubusercontent.com/30960668/145350937-c3159a14-c91b-4a13-9483-172b31bcbdcf.png)

`Piccolo Admin` and request directly in browser remain unchanged.

@gmos Can you please check this on your local `Piccolo Api` installation and confirm that everything works.?

**Comment by dantownsend** [comment]
@sinisaos That's some good investigative work - thanks.

When we make this change we'll also have to update Piccolo Admin at the same time.

**Comment by dantownsend** [comment]
@sinisaos Ah, just realised you managed to make it compatible with the current Piccolo Admin, great.

**Comment by sinisaos** [comment]
@dantownsend Yes, Piccolo Admin works. If you want I can do PR.

**Comment by dantownsend** [comment]
@sinisaos That would be great, thanks

**Comment by dantownsend** [comment]
I've released it to PyPI now - version 0.29.2.

Seems to fix the problem. Thanks both.
